pkg/server/backend_manager.go

```diff
@@ -57,21 +60,66 @@ func newBackend(conn agent.AgentService_ConnectServer) *backend {
 	return &backend{conn: conn}
 }
 
-// BackendManager is an interface to manage backend connections, i.e.,
-// connection to the proxy agents.
-type BackendManager interface {
-	// Backend returns a single backend.
-	Backend() (Backend, error)
+// BackendStorage is an interface to manage the storage of the backend
+// connections, i.e., get, add and remove
+type BackendStorage interface {
+	// GetBackend returns a backend associating to the agentID.
+	GetBackend(agentID string) Backend
 	// AddBackend adds a backend.
 	AddBackend(agentID string, conn agent.AgentService_ConnectServer) Backend
 	// RemoveBackend removes a backend.
 	RemoveBackend(agentID string, conn agent.AgentService_ConnectServer)
+	// NumBackends returns the number of backends.
+	NumBackends() int
+}
+
+// BackendManager is an interface to manage backend connections, i.e.,
+// connection to the proxy agents.
+type BackendManager interface {
+	// Backend returns a single backend.
+	Backend(context.Context) (Backend, error)
+	BackendStorage
 }
 
 var _ BackendManager = &DefaultBackendManager{}
 
 // DefaultBackendManager is the default backend manager.
 type DefaultBackendManager struct {
+	*DefaultBackendStorage
+}
+
+func (dbm *DefaultBackendManager) Backend(_ context.Context) (Backend, error) {
+	be := dbm.DefaultBackendStorage.GetRandomBackend()
+	if be == nil {
+		return nil, &ErrNotFound{}
+	}
+	return be, nil
+}
+
+var _BackendManager = &DestIPBackendManager{}
+
+type DestIPBackendManager struct {
+	BackendStorage
+}
+
+func (dbm *DestIPBackendManager) Backend(ctx context.Context) (Backend, error) {
+	destIPInf := ctx.Value(DestIPContextKey)
+	if destIPInf == nil {
+		return nil, errors.New("no agentID found in the context")
+	}
+	agentID, ok := destIPInf.(string)
+	if !ok {
+		return nil, errors.New("type assertion failed")
+	}
+	be := dbm.BackendStorage.GetBackend(agentID)
+	if be == nil {
+		return nil, &ErrNotFound{}
```

Comment:
This does not seem desirable. If the agent is down on the relevant node, we are saying that we will refuse to attempt to send traffic to that node. Its possible the node is up and that we are exec a request to fix the agent. The existing kube-dns code should route the traffic. I think we want this as a best effort to route through the best tunnel rather than only routing through that tunnel.

Reply:
If you do have a use case for *only* sending to the requested IP, I would suggest we have an addition parameter which is a ?strictness or ?exactDestination flag. That way the admin can control which way it goes.